Build Python package from source distributions with Meson

jared321 opened the pull request:
The work on this branch is finished and represents work that should hopefully make it into `main`.

My self-review notes

- [x] Careful review of all changes made here.
- [x] Confirm that C++ build script is building and installing CLTs correctly.
  * Used clang++ v15.0.0 with MPICH v4.2.3 and Open MPI v5.0.6 both installed by homebrew.
  * Meson v1.6.0 installed from homebrew and installed manually in dedicated venv as described in README.
  * Confirmed that Eigen v3.4.0 is installed automatically when eigen not in system
  * Confirmed that Eigen installed via homebrew was found and used by Meson so long as `pkgconf` was also installed by homebrew to provide `pkg-config`
  * `otool -L` on installed CLTs looks good and on intermediate library tests as well.
  * Library and headers installed as expected
  * Can run tests from `builddir` with `meson test` and `meson test <test name>` (e.g., with `crn` and `tree`)
  * Compiler flags reasonable and different for both release and debug builds
  * `-DNDEBUG`, `-D_LIBCPP_ENABLE_ASSERTIONS`, `-DSILENT`, and `-D_USEMPI` appear when appropriate only
- [x] Confirm that developers can use meson to build the CLTs directly as debug builds with different warning levels and with warnings treated as errors.
- [x] Confirm that shell procedures in README work as expected.
  * Inspected manually built source distributions to confirm correct contents.
  * Confirmed that LICENSE file is included in the distribution and in the installation.
  * Confirmed that installation had minimal, correct content, which includes the CLTs but not the library or headers.
  * Ran `otool -L` on installed CLTs and all looks good.
  * Confirmed that LICENSE file was installed.
  * Confirmed that developer-mode installation was in expected location and had minimal, correct content (more than actual installation).
  * Ran `otool -L` on developer-mode installed CLTs and all looks good.
- [x] Install an older version of Meson and confirm that a useful message is reported to user that indicates the minimum version that is required.
  * Installed v1.5.2 in dedicated venv as described in README
  * Meson prints useful error message which build from sdist, installing in developer mode, and when using build script.
  * Non-zero exit code returned in all cases.
- [x] Confirm that pip install does not print meson setup, compile, or install output by default, but that these can be viewed by passing `-v` to pip install
- [x] Break setup.py in multiple ways and confirm that `pip install` fails with a useful message.
  * Typo in meson command triggered failure with both stdout/stderr printed with pip install in non-verbose mode. 
  * Inability to find MPI installation triggered failure with both stdout/stderr printed with pip install in non-verbose mode. 
- [x] Setup several different (compiler suite, MPI implementation) setups on a cluster using LMod, install Python package, and run tests.
  * On LCRC improv cluster (Linux) with Python v3.11.6 & ninja v1.12.1 installed by modules; Meson v1.6.0, installed in dedicated venv as indicated in README.
  * For each setup, built CLTS+library+tests in release and debug.  Checked Meson output and compiler/linker flags.  Ran tests.  Build/installed Python package with verbose `pip install`.  Checked Meson output and compiler/linker flags (temporarily added `-v` to meson compile command in `setup.py`) and confirmed that tests pass.
  * Confirmed two MPI processes running during Python package test execution using `top`
  * Tested GCC v13.2.0 with Open MPI v5.0.2 (built with GCC v13.2.0) loaded by modules.  No issues and tests passed.
  * Tested GCC v13.2.0 with Open MPI v4.2.0 (built with GCC v13.2.0) loaded by modules.   Failed potentially due to bug fixed at v4.3.0, which is not installed (https://github.com/pmodels/mpich/issues/7036).
  * Tested Intel OneAPI v2023.2.1 with Open MPI v4.1.6 (built with OneAPI v2023.2.1) loaded by modules.  No issues and tests passed.
  * Tested Intel OneAPI v2023.2.1 with Intel MPI v2021.10.0 (built with OneAPI v2023.2.1) loaded by modules.  Note that Meson finds `mpic++`, which is installed by Intel but curiously wraps g++.  The correct wrapper is `mpiicpc`.  As in all other cases, however, the wrappers are not called.  Instead the correct Intel compiler is called with compatible flags.  Indeed, the flags seem to be taken from the `impi` pkg-config settings.  This ran the first test for a very long time and I only saw one MPI process running (at 100%) during that time.  However, I think that only one process is running at times for that test.  Not clear what the issue is yet.  Reran from a clean login successfully.  I have seen some issues with the OpenMPI modules still having an effect on the env vars after unloading and despite loading a new MPI installation.  I wonder if what I saw previously was the effect of a bad environment.
- [x] Remove MPI installations in all GitHub Actions to confirm failure and therefore determine that Meson is finding our MPI installation rather than a default MPI installation.
- [x] Careful review of all log output of GitHub Actions to confirm correctness
- [x] Download and inspect coverage reports generated by Action.
- [x] Review README as rendered by GH, confirm correct content, and confirm that links work and are correct.
- [x] Confirm that all actions are passing

Adding @asemposki as a reviewer.  Could you please review
* the README for correctness, clarity, and usefulness,
* the README to determine if the new installation process is acceptable for Taweret and to determine how to update the Taweret documentation in accord with the new process, and
* the set of GH actions to determine if they are sufficient.

You might want to review `tools/build_openbt_clt.sh` and `openbtmixing_pypkg/setup.py` to understand how `meson` is used and integrated into this project.

jared321 committed:
(Issue #6) Try changing configure/build system to meson.

jared321 committed:
(Issue #6) Remove autotools files

jared321 committed:
(Issue #6) This gives me a working installed wheel.

Note that I am building the wheel and using it on the same machine.  Therefore,
the fact that the package's tests are passing does not imply that they will
pass if installed on a different machine.  For instance, a different machine
might have OpenMPI installed in a different location.  Note that not all
C++-based tests are building or passing.

jared321 committed:
(Issue #6) Update CLT and Python package scripts.

Ideally we will be able to use these for local builds and in GH actions.  Note
that the wheel script is presently just for macOS builds.

jared321 committed:
(Issue #6) Ask action to build/test wheels with CLTs inside.

jared321 committed:
(Issue #6) Clean docs.

jared321 committed:
(Issue #6) Add back files needed to avoid merge conflict with main.

I need to resolve the conflict so that GH actions run.

jared321 committed:
(Issue #6) Try installing ninja+meson

jared321 committed:
(Issue #6) Stick to open-mpi for now

jared321 committed:
(Issue #6) Install delocate/auditwheel.

jared321 committed:
(Issue #6) Clean GH action

jared321 committed:
(Issue #6) Archive wheel

jared321 committed:
(Issue #6) Get automatic builds when installing from sdist.

This is now using meson under the hood to identify the local MPI installation
and use this to build the C++ OpenBT CLTs.  These are then included in a
locally-built binary wheel and installed.

Ran the in-package tests and these are all passing.  I see four MPI processes
running at 100%.

jared321 committed:
(Issue #6) Add in meson options.  Build in release for Py Pkg.

jared321 committed:
(Issue #6) Remove old releases.

jared321 committed:
(Issue #6) Add in more TODOs.

jared321 committed:
(Issue #6) CLTs should be standalone.

jared321 committed:
(Issue #6) No need for library dependence hacking now.

Python package unittests passing and I don't see libopenbtmixing as an external
dependence anymore.

jared321 committed:
(Issue #6) Try fixing GH action.

jared321 committed:
(Issue #6) Get debug info

jared321 committed:
(Issue #6) Still getting more info

jared321 committed:
(Isssue #6) Try installing meson via homebrew

jared321 committed:
(Issue #6) Try coverage and get more debug info.

jared321 committed:
(Issue #6) Use verbose flag correctly with meson.

jared321 committed:
(Issue #6) Try running on Ubuntu.

jared321 committed:
(Issue #6) Still working on Ubuntu.

jared321 committed:
(Issue #6) Bet this doesn't make a difference.

jared321 committed:
(Issue #6) Newer Ubuntu/newer meson?

jared321 committed:
(Issue #6) Try OpenMPI

jared321 committed:
(Issue #6) Try MPICH again so that I can see if it's in pkg-config.

jared321 committed:
(Issue #6) It looks like mpich should be in pkg-config

jared321 committed:
(Issue #6) More information?

jared321 committed:
(Issue #6) meson looked for ompi-cxx in pkg-config with only MPI.

jared321 committed:
(Issue #6) Go back to OpenMPI.

Even with the last change, meson was still looking for ompi-cxx.  Not clear why
it didn't find and use the mpi* wrappers made available via PATH.

jared321 committed:
(Issue #6) Temporary hack to get OpenMPI working in GH actions.

jared321 committed:
(Issue #6) Worked with 24.04 with OpenMPI hack.  Try others.

jared321 committed:
(Issue #6) Install latest meson into test venv from PyPI.

The meson documentation suggests that this should be preferred over installation
via package manager since the rapid development of meson will likley result in
only significantly old versions available through the managers.  This is the
case for Ubuntu 20.04 and 22.04 at the very least.  Also, it looks like meson
might now correctly find MPICH installations as of the latest versions, v1.6.0,
which is too new even for Ubuntu 24.04.

jared321 committed:
(Issue #6) Keep meson in user space?

jared321 committed:
(Issue #6) Install meson in its own venv and in PATH?

jared321 committed:
(Issue #6) Concentrate on Ubuntu 24.04.

We had tests pass at commit 9b6049b9de with this OS when meson was installed by
apt at version 1.3.2-1ubuntu1.  See if we can repeat that success but with a
locally installed meson.

jared321 committed:
(Issue #6) Try older meson version.

jared321 committed:
(Issue #6) See if passing as well on older Ubuntus with meson 1.3.2.

jared321 committed:
(Issue #6) Try using modern meson versions.

Beginning at v1.4.0, meson was enabling C++ asserts by default.  It appears that
these are what are causing the failures.  Turn asserts off to see if we can run
with the most modern meson versions.

jared321 committed:
(Issue #6) Stick with v1.6.0 and greater.  Don't require MPI implementation.

This should allow users to build the Python package using whatever MPI
implementation they expose to meson.  Have not tested this yet.

jared321 committed:
(Issue #6) Try building with MPICH.

jared321 committed:
(Issue #6) Remove setup that failed with MPICH.

The remaining Ubuntu setups passed, but I need to see if the two macOS setups
will pass.

jared321 committed:
(Issue #6) Test C++ & PyPkg separately.  Try more setups.

jared321 committed:
(Issue #6) Allow for debug builds & do work more cleanly.

jared321 committed:
(Issue #6) Lower N MPI processes to get OpenMPI tests working in GH actions.

jared321 committed:
(Issue #6) Try testing more installation/use methods.

jared321 committed:
(Issue #6) Archive coverage results and start cleaning.

jared321 committed:
(Issue #6) Fix copy/paste error.

jared321 committed:
(Issue #6) Continue cleaning GC actions.

jared321 committed:
(Issue #6) See if we can install anaconda

jared321 committed:
(Issue #6) Try fixing yml error.

jared321 committed:
(Issue #6) Fix next yml error.

jared321 committed:
(Issue #6) Try installing MPI via mpi4py

jared321 committed:
(Issue #6) See if conda can build package.

jared321 committed:
(Issue #6) Try installing build via pip.

jared321 committed:
(Issue #6) Try using more MPI implementations.

jared321 committed:
(Issue #6) Try using the requested MPI implementation.

jared321 committed:
(Issue #6) conda couldn't install Intel MPI.

jared321 committed:
(Issue #6) Try macOS and skip MPICH for now.

jared321 committed:
(Issue #6) Try more OSs with miniconda.

jared321 committed:
(Issue #6) Get meson via brew & see if we can skip mpi4py install.

jared321 committed:
(Issue #6) Try conda MPICH with macOS.

jared321 committed:
(Issue #6) MPICH not working with macOS or Ubuntu through conda.

jared321 committed:
(Issue #6) Linux CLT installation needs library location set.

The macOS test binaries have the OpenBT library apparently hardcoded in (based
on otool -l).  However, this path is not set as the RPATH in the Linux builds
(based on ldd).

jared321 committed:
(Issue #6) Get more info for debugging.

jared321 committed:
(Issue #6) Try getting yet more info.

jared321 committed:
(Issue #6) See if this gets tests passing on Ubuntu.

jared321 committed:
(Issue #6) Continue to build out and clean.

jared321 committed:
(Issue #6) Continue building out and cleaning.

jared321 committed:
(Issue #6) Clean-up file organization and meson build file.

jared321 committed:
(Issue #6) Clean-up.  Only build the bare minimum for Python package.

jared321 committed:
(Issue #6) First draft of new installation instructions.

jared321 committed:
Merge branch 'main' into 6_MesonProto

jared321 committed:
(Issue #6) Remove files needed by autotools scheme.

jared321 committed:
(Issue #11) Return non-zero exit codes on error.

jared321 committed:
(Issue #6) Cleaning as part of PR review.

I will also try to investigate if MPICH fails with Ubuntu24 for all Python
versions and hope to get an idea why it fails.

jared321 committed:
(Issue #6) Try to fix GH action file.

jared321 committed:
(Issue #6) Try widdling down set of test setups.

jared321 committed:
(Issue #6) Try single test.

jared321 committed:
(Issue #6) Get MPI information.

jared321 committed:
(Issue #6) See if this gets Ubuntu+MPICH working.

jared321 committed:
(Issue #6) Get more MPICH info

jared321 committed:
(Issue #6) Don't install MPICH to see if there is a default installation.

jared321 committed:
(Issue #6) Just print whiches.

jared321 committed:
(Issue #6) See if Meson fails.

jared321 committed:
(Issue #6) Try installing a suggested dev package for 24.04

jared321 committed:
(Issue #6) Reset sdist GH action.

I will leave 24.04+MPICH tests excluded for now.

jared321 committed:
(Issue #6) More cleaning as part of PR review.

jared321 committed:
(Issue #6) Cleaning as part of PR review.

jared321 committed:
(Issue #6) Clean-up meson build config as part of PR review.

jared321 committed:
(Issue #6) Cleaning up as part of PR review.

jared321 committed:
(Issue #6) More cleaning of README for review.

jared321 committed:
(Issue #6) Expand build script's configurability and abilities.

For macOS, the hombrew installation of Eigen has it available as "eigen3" in
pkg-config.  Therefore, update the dependency in Meson.

jared321 committed:
(Issue #6) Add more content to README.

jared321 committed:
(Issue #6) Don't install MPI in macOS-13 GH test actions.

I hope that I am intentionally breaking these tests by doing this.  In
particular, I want to confirm that under normal operations Meson isn't finding a
default MPI installation I don't know about.  The effected tests should fail
when Meson notifies that it can't satisfy the MPI dependence.

I will do the same for all different OSs used nominally in each test action.

jared321 committed:
(Issue #6) Test no MPI with macOS 14.

jared321 committed:
(Issue #6) Test no MPI with macOS-15.

jared321 committed:
(Issue #6) No MPI with Ubuntu 20.04

jared321 committed:
(Issue #6) No MPI with Ubuntu 22.04

jared321 committed:
(Issue #6) No MPI with Ubuntu 24.04

jared321 committed:
(Issue #6) Reset GH test actions to full and correct setups.

All intentional breaks did indeed break as expected.

jared321 committed:
(Issue #6) Show compile/link commands in verbose pip install mode.

By default, pip install is non-verbose.  If a user/developer explicitly asks for
verbose mode, they probably want to see or won't mind seeing the commands.  This
will also improve the utility of the GH action logging, which does verbose pip
installs.

jared321 committed:
Add useful warning

jared321 commented:
@jcyannotty Alexandra and I can both build & install from source on this branch.  The tests are passing as well.  Could you please review and test?

Note that with this installed alongside Taweret, the Taweret tests failed.  I presently believe (TBC) that the state of the OpenBT code in this branch is too new for Taweret's `main` branch, but too old for your updated architecture.  Is this possible?

jcyannotty commented:
> @jcyannotty Alexandra and I can both build & install from source on this branch. The tests are passing as well. Could you please review and test?
> 
> Note that with this installed alongside Taweret, the Taweret tests failed. I presently believe (TBC) that the state of the OpenBT code in this branch is too new for Taweret's `main` branch, but too old for your updated architecture. Is this possible?

I will review sometime this week. The code on this repository should work with the code on 95_TreesUpdate on the Taweret repo. The unit tests on the OpenBT repo are the exact same as in Taweret (at least I don't remember making any changes)

asemposki commented:
My review: 

- Reviewed the README for clarity and usefulness, and performed the steps locally to ensure the package builds properly;
- Reviewed the GH actions---these look sufficient to me, and cover all of the tests that I can think of currently;
- Once we've ironed out the Taweret compatibility and gotten this working (hopefully with @jcyannotty's updates to Taweret on the 95_TreesUpdate branch), I will check the Taweret documentation and change it to add this process in

jared321 commented:
As a final check, I tried to run the Taweret tests using the openbtmixing package installed from this PR.  It fails with the following message
```
Cannot find openbt executables.Please specify the path using the argumentlocal_openbt_path in the constructor.
```

When I switched to `95_TreesUpdate`, I had to change in `trees.py` the line
```
from openbtmixing import OpenbtMix
```
to
```
from openbtmixing import Openbtmix
```
After that, all but one test passed.  The one failure is in `test_mixing`
```
>       assert fit["pbd"] == 0.7, "check prob of birth and death"
E       KeyError: 'pbd'

tests/test_trees.py:87: KeyError
```

jared321 commented:
@jcyannotty Since we are getting closer to merging this, are there "official release" duties that need to be done on this branch before merging?  For instance, currently v1.0.1 is in PyPI.  Does this qualify as a newer version?

jcyannotty commented:
> @jcyannotty Since we are getting closer to merging this, are there "official release" duties that need to be done on this branch before merging? For instance, currently v1.0.1 is in PyPI. Does this qualify as a newer version?

I would say yes this qualifies as a new version. So V1.1.0 now?

jcyannotty commented:
> If it's version v1.1.0, then I understand that these changes are not a major version change, but just a minor change. In particular, people shouldn't have to alter their pre-existing code significantly because the package's interface has changed. There are simply some new tweaks, fixes, or additions.
> 
> If that's correct, then v1.1.0 sounds good to me!

Yes, in that case v1.1.0 should be appropriate.

jared321 committed:
(Issue #6) These changes are to be considered as a minor update.

jared321 committed:
(Issue #6) Ubuntu 20.04 GH action runner being deprecated.

jared321 committed:
(Issue #6) Try getting anaconda GH action passing again.

jared321 committed:
(Issue #6) Cleaning anaconda GH action.

jared321 committed:
(Issue #6) Branch is also an update to the C++ portion of repo.

Therefore, update the C++ version in meson in the same way that the Python
package's version was updated.

jared321 committed:
(Issue #6) Point meson toward LICENSE file.

jared321 committed:
(Issue #6) Need LICENSE file in correct place for sdist.

jared321 committed:
(Issue #6) Include extra LICENSE file in sdist.

jared321 commented:
@jcyannotty I updated the version of the Python package (in `VERSION`) as well as the version of the C++ code (in `meson.build`).  I also had to modernize some of the actions.  But now, all actions in your repo are passing.  Also, all actions in Taweret are passing using the code from the latest commit in this branch.

When you are ready, you can merge this branch into `main` and publish the wheels on PyPI (in whatever order you desire).  Let me know when they're up and I'll reset the Taweret actions to install your package from PyPI.

P.S. You might want to review Issue #12 first.

jcyannotty committed:
Update LICENSE

jcyannotty committed:
Update setup.py for new license format

jcyannotty committed:
Add lines in test_py_devmode.yml to install meson python module

jcyannotty committed:
Install meson module for mac os

jcyannotty committed:
Rm the brew install meson command and rely solely on pip

jcyannotty committed:
Debug the mac os meson

jcyannotty committed:
More debugging

jcyannotty committed:
Debug try with pipx

jcyannotty committed:
Try virtual environment for macOS
